### PR TITLE
Pure python agent functions

### PR DESCRIPTION
**Pull request by mondus**
PR to add support for pure python agent functions. E.g. Agent functions written in python which are the source to source compiled to FLAME GPU C++ code for processing with the runtime compilation process.

 - [x] Add the code generator module to the build process so it is included in the generated python package
 - [x] Add the code generator tests
 - [x] Add an example for circles or boids model
 - [x] Add some documentation and clarify that this is an **experimental** feature

Known limitations:
- Lots of python features do not translate to cpp. Tough luck!
- Error handing is poor due to lack of helpful error reporting in jitify. A jitify extension to support querying of errors would allow the cpp error to be remapped to the original python source.
- Agent function conditions are not supported (but probably could be)
- You cant generate device functions that use arguments by reference (everything is assumed to be pass by value). There is no pythonic way to distinguish.

### CODE REVIEW

File: swig/python/codegen/codegen.py [hunk 1]
```diff
@@ -0,0 +1,1030 @@
...
+                  "float64": "double"
+                  }
+    
+    # getVariableType and setVariableType functions are added dynamically    
+    fgpu_funcs = [ "getID", "getStepCounter", "getThreadIndex" ]   
+    fgpu_attrs = ["ALIVE", "DEAD"]
+    fgpu_input_msg_funcs = ["getIndex"] 
+    fgpu_output_msg_funcs = []
+    fgpu_agent_out_msg_funcs = ["getID"]
+    fgpu_env_funcs = ["containsProperty"] # TODO: Get macro property
```
Comment (Robadob) [inline swig/python/codegen/codegen.py:87]: TODO

**Review by Robadob — COMMENTED** [review]
Looks OK to be merged, given it's largely standalone so shouldn't upset anything else (so long as the CMake is stable, @ptheywood should test this on Linux when he's back).

Spotted a TODO in the code.

Not looking forwards to maintaining this as the API expands (e.g. graphs)

File: swig/python/codegen/codegen.py (hunk 1 above)
Comment (mondus) [reply swig/python/codegen/codegen.py:87]: Macro properties are a feature that it maybe quite hard to support without considering much broader support for array types. I think it would be better to merge with this as a known limitation at this stage.

**Comment by mondus** [comment]
Maintenance should not be too bad. There is a list of permitted device functions that will need to keep parity.

**Comment by mondus** [comment]
Macro Environment properties now supported and tested. It doesn't actually require array support only support for `ast.SubScript`.

**Comment by mondus** [comment]
Now has support for pure python agent and device functions. This is now properly pure and not just strings of python functions. Things I don't like. The singleton used by the python device functions is pyflamegpu obviously the modules does not have the device functions within it which would break syntax checkers.

I don't recommend any other feature additions at this stage.

File: swig/python/CMakeLists.txt [hunk 2]
```diff
@@ -218,6 +230,18 @@ add_custom_command(TARGET ${PYTHON_MODULE_NAME} POST_BUILD     # Adds a post-bui
...
         "${PYTHON_LIB_OUTPUT_DIRECTORY}/${PYTHON_MODULE_NAME}/include" # <--this is out-dir path
     ) 
+
+# Create code generator directory
+add_custom_command(TARGET ${PYTHON_MODULE_NAME} POST_BUILD 
+    COMMAND ${CMAKE_COMMAND} -E make_directory "${PYTHON_LIB_OUTPUT_DIRECTORY}/${PYTHON_MODULE_NAME}/codegen"
+)
+# Copy python codegen source files to library output directory ready for packaging
+foreach( file_py ${PYTHON_CODEGEN_SRC_FILES})
+  add_custom_command(TARGET ${PYTHON_MODULE_NAME} POST_BUILD
```
Comment (ptheywood) [inline swig/python/CMakeLists.txt:240]: This does not behave correctly for incremental builds where only parts of `codegen` are modified.

Under linux (and presumably windows too), (probably) because this is a `POST_BUILD` event on the `pyflamegpu` target (`PYTHON_MODULE_NAME`) the files are only copied when the pyflamegpu target has been built. 

I.e. if the pyflamegpu module is up to date so it does not need to be built, modifications to the codegen are not copied across. 

1. Build `pyflamegpu`
2. Edit `swig/pythong/codegen/codegen.py`
3. Invoke a build (not a clean and build / rebuild in VS terms)
4. Check the codegen file in the venv (`lib/<build_type>/python/venv/lib, it will not have been updated. (i.e. `lib/Release/python/pyflamegpu/codegen/codegen.py`)


Instead, iff a clean build is done, or a change which would invoke an incremental build of pyflamegpu is triggered the file is updated (and so is the version in the venv)

This seems to suggest that the `DEPENDS` statement below is insufficient, though i'm not sure what the correct way to fix this is off the top of my head. 

It could also require an `OUTPUTS` section adding to this custom command based on a quick read of the [docs](https://cmake.org/cmake/help/v3.0/command/add_custom_command.html), but I have a feeling its more to do with the POST_BUILD not being triggered (Editing `pyflamegpu.i` and triggering a build does work). 

This might be a fairly involved restrucutre to get this to trigger without a pyflamegpu.i rebuild (may need an intermediate target to avoid the long compile step, but this is speculation).

File: swig/python/CMakeLists.txt [hunk 3]
```diff
@@ -261,13 +285,14 @@ if (VISUALISATION)
...
 # Configure and build the python package
-# The file configurations above will already have output setup.py and __init__.py to the appropriate places and created the directory for the module
+# The file configurations above will already have output setup.py, __init__.py and the code generator module files to the appropriate places and created the directory for the module
 add_custom_command(TARGET ${PYTHON_MODULE_NAME} POST_BUILD
 	# copy generated python file to library output location
 	COMMAND ${CMAKE_COMMAND} -E copy ${PYTHON_LIB_TEMP_DIRECTORY}/${PYTHON_MODULE_NAME}/${PYTHON_MODULE_NAME}.py ${PYTHON_MODULE_NAME}
 	# copy the compiled pyd file to library output location
 	COMMAND ${CMAKE_COMMAND} -E copy $<TARGET_FILE:${PYTHON_MODULE_NAME}> ${PYTHON_MODULE_NAME}
 	COMMAND ${Python3_EXECUTABLE} setup.py bdist_wheel
+  DEPENDS ${PYTHON_CODEGEN_SRC_FILES}
```
Comment (ptheywood) [inline swig/python/CMakeLists.txt:295]: This may also be insufficient as above, due to it being a post_build event so it doesn't always fire. Not 100% though.

File: swig/python/codegen/__init__.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
+# coding: utf-8
+from six.moves import cStringIO
```
Comment (ptheywood) [inline swig/python/codegen/__init__.py:2]: We should not need to use `six`, a python 2 / 3 compatibiliity library, given we explicitly only support python 3.7+

You should use [`io.StringIO`](https://docs.python.org/3/library/io.html#io.StringIO) instead (based on the `six` docs).

File: swig/python/setup.py.in [hunk 5]
```diff
@@ -35,6 +35,9 @@ setup(
   'Topic :: Scientific/Engineering',
   ],
   package_data={
-  '@PYTHON_MODULE_NAME@':['$<TARGET_FILE_NAME:@PYTHON_MODULE_NAME@>', @FLAMEGPU_INCLUDE_CLEAN@@FLAMEGPU_PYTHON_PACKAGE_DATA_OS_SPECIFIC@],
+  '@PYTHON_MODULE_NAME@':['$<TARGET_FILE_NAME:@PYTHON_MODULE_NAME@>', @FLAMEGPU_CODEGEN_INCLUDE_CLEAN@@FLAMEGPU_INCLUDE_CLEAN@@FLAMEGPU_PYTHON_PACKAGE_DATA_OS_SPECIFIC@],
   },
+  install_requires=[
+    'six', 'astpretty',
```
Comment (ptheywood) [inline swig/python/setup.py.in:41]: `six` should not be a required dependency as we do not care for python 2 compatibility. See previous comments.

File: swig/python/codegen/codegen.py [hunk 6]
```diff
@@ -0,0 +1,1172 @@
+from __future__ import print_function, unicode_literals
+import six
```
Comment (ptheywood) [inline swig/python/codegen/codegen.py:2]: `six` should not be a dependency, python 2 compatibility is not required.

File: swig/python/codegen/codegen.py [hunk 7]
```diff
@@ -0,0 +1,1172 @@
...
+import sys
+import ast
+import os
+import tokenize
+import warnings
+from six import StringIO
```
Comment (ptheywood) [inline swig/python/codegen/codegen.py:8]: Use `io.StringIO` instead, which is core python3.

File: swig/python/codegen/codegen.py [hunk 8]
```diff
@@ -0,0 +1,1172 @@
...
+        return cpp_func_name
+              
+
+    def fill(self, text = ""):
+        "Indent a piece of text, according to the current indentation level"
+        self.f.write("\n"+"    "*self._indent + text)
+
+    def write(self, text):
+        "Append a piece of text to the current line."
+        self.f.write(six.text_type(text))
```
Comment (ptheywood) [inline swig/python/codegen/codegen.py:174]:
```suggestion
        self.f.write(str(text))
```

`six.text_type` is just `str` in python3.

File: swig/python/codegen/codegen.py [hunk 9]
```diff
@@ -0,0 +1,1172 @@
...
+            self._CallArguments(t.value)
+            self.write(")")
+            # write outer function (call args will be completed by _Call)
+            self.write(f".{t.attr}")
+            
+   
+        else:
+            self.RaiseError(t, "Unsupported function call syntax")
+     
+
```
Comment (ptheywood) [inline swig/python/codegen/codegen.py:545]: There are unneccesary blank lines that could be cleaned up, though arguably we don't have python linting rules set up but maybe we should now we have more python (see https://github.com/FLAMEGPU/FLAMEGPU2/issues/666)

File: swig/python/codegen/codegen.py [hunk 10]
```diff
@@ -0,0 +1,1172 @@
...
+    except StopIteration:
+        pass
+    else:
+        for x in seq:
+            inter()
+            f(x)
+
+class CodeGenException(Exception):
+    """ Generic exception for errors raised in code generation """
+    pass
```
Comment (ptheywood) [inline swig/python/codegen/codegen.py:26]: This `pass` statement is not actually necessary, as docstrings count as an expression (they are not a comment, as they are become a member of the class). 

No harm in keeping it, just something odd that I found out recently.

**Review by MILeach — CHANGES_REQUESTED** [review]
Couple of comments but nothing major, looks good and tests are passing fine for me on Ubuntu. CMake stuff looks OK to me but it's probably worth Pete taking a look too!

File: swig/python/codegen/__init__.py [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+        return codegen(tree)
+    elif isinstance(function, Callable):
+        # If a Callable has been passed directly then we need to seek any functions with the 'pyflamegpou.device_function' decorator to include in the source for translation
+        # There is no need to use unwrap on `function` as it uses `functools.wrap`
+        # get the module of the function
+        module = inspect.getmodule(function)
+        # get functions
+        module_functions = inspect.getmembers(module, inspect.isfunction)
+        prepend_source = ""
+        if len(module_functions):
```
[the remark this answers is not in the document]
Comment (MILeach) [reply swig/python/codegen/__init__.py:32]: This comment is with reference to the if statement on line 32

File: swig/python/codegen/codegen.py [hunk 12]
```diff
@@ -0,0 +1,1172 @@
...
+    is disregarded. """
+
+    # represents built in functions
+    pythonbuiltins = ["abs", "float", "int"]
+    
+    # basic types
+    basic_arg_types = ['float', 'int']
+    
+    # supported math constansts    
+    mathconsts = {"pi": "M_PI",
```
Comment (ptheywood) [inline swig/python/codegen/codegen.py:40]: Why is the other python `math` constatant `math.tau` not supported? ([docs](https://docs.python.org/3/library/math.html#math.tau)). 

It's available as python 3.6, which is currently our minimum but moving to >= 3.7 due to 3.6 being EOL, so it should be safe to include this.

**Review by ptheywood — CHANGES_REQUESTED** [review]
Incremental builds do not behave as they should, when only ediitng the codegen/* files, as they are only copied by POST_BUILD events on the pyflamegpu target. So if `pyflamegpu` is up to date, they are not copied into the wheel or re-installed into the venv. The pyflamegpu target needs to depend on the files, but ideally not trigger a rebuild of `pyflamegpu.i` which takes a decent  length of time to compile. 

There is an absence of documentation within this repository on the presnece of this feature, and it's experimental state. There should be atleast something in the readme. 

There are also no tests which check that the device code actually behaves as intended as far as I can tell, the tests only check the ast side of things. This is fine for the new unit's but doesn't cover the integration. Relying on the main (python and RTC tests) is fine for most features, but there should be atleast one test which invokes device code from the pure python. Ideally with decent coverage, but realistically just covering edge cases would be worthwhile. 


I've left a number of comments throughout too.

Otherwise it does run and the 48 tests pass under linux, release with seatbelts.

History will want cleaning up prior to merge too.

File: swig/python/CMakeLists.txt (hunk 3 above)
Comment (mondus) [reply swig/python/CMakeLists.txt:295]: 80% of the time it works every time!

File: swig/python/codegen/codegen.py (hunk 10 above)
Comment (mondus) [reply swig/python/codegen/codegen.py:26]: I think ast parsing dumps them though

File: tests/swig/python/codegen/test_codegen.py [hunk 13]
```diff
@@ -0,0 +1,817 @@
...
+    def test_for_else(self):
+        self._checkException(py_for_else, "For else not supported")
+        
+    def test_for_range(self):
+        # use of range based for loops (calling range with different number of arguments)
+        self._checkExpected(py_for_range_arg1, cpp_for_range_arg1)
+        self._checkExpected(py_for_range_arg2, cpp_for_range_arg2)
+        self._checkExpected(py_for_range_arg3, cpp_for_range_arg3)   
+        # check that non range function loops are rejected
+        self._checkException(py_for_unsupported, "Range based for loops only support")
```
[the remark this answers is not in the document]
Comment (mondus) [reply tests/swig/python/codegen/test_codegen.py:514]: The full error string is actually much longer (and may include variable names etc) but the `_checkException` function tests that the error string begins with the provided text. Have extended the test will more of the error string regardless.

File: tests/swig/python/codegen/test_codegen.py [hunk 14]
```diff
@@ -0,0 +1,817 @@
...
+        self._checkExpected("-1.**8", "(-pow(1.0, 8));")
+        self._checkExpected("not True or False", "((!true) || false);")
+        self._checkExpected("True or not False", "(true || (!false));")
+
+    def test_integer_parens(self):
+        self._checkException("3 .__abs__()", "Unsupported") # should resolve to unsupported function call syntax
+
+    def test_huge_float(self):
+        self._checkExpected("1e1000", "inf;")
+        #self._checkExpected("-1e1000")
```
[the remark this answers is not in the document]
Comment (mondus) [reply tests/swig/python/codegen/test_codegen.py:533]: yes

File: tests/swig/python/codegen/test_codegen.py [hunk 15]
```diff
@@ -0,0 +1,817 @@
...
+    def test_lambda_parentheses(self):
+        self._checkException("(lambda: int)()", "Lambda is not supported")
+
+    def test_chained_comparisons(self):
+        self._checkExpected("1 < 4 <= 5", "1 < 4 <= 5;")
+        self._checkExpected("a is b is c is not d", "a == b == c != d;")
+
+    def test_function_arguments(self):
+        # only flame gpu functions or device functions are supported
+        self._checkException("def f(): pass", "Function definitions require a")
```
[the remark this answers is not in the document]
Comment (mondus) [reply tests/swig/python/codegen/test_codegen.py:560]: See comment above. The actual exception is a complete message. The test against it is just checks the start if the exception string to make sure that a relevant exception is thrown.

File: tests/swig/python/codegen/test_codegen.py [hunk 16]
```diff
@@ -0,0 +1,817 @@
...
+    # device functions, arg types and calling
+    
+    def test_fgpu_device_func(self):
+        self._checkExpected(py_fgpu_device_func_args, cpp_fgpu_device_func_args)
+        # function argument requires a type
+        self._checkException(py_fgpu_device_func_no_type, "Device function argument requires type annotation")
+        # no function return type should use void
+        self._checkExpected(py_fgpu_device_func_no_return_type, cpp_fgpu_device_func_no_return_type)
+        # function requires decorator
+        self._checkException(py_fgpu_device_func_no_decorator, "Function definitions require a single pyflamegpu decorator of")
```
[the remark this answers is not in the document]
Comment (mondus) [reply tests/swig/python/codegen/test_codegen.py:811]: See comment above. The actual exception is a complete message. The test against it is just checks the start if the exception string to make sure that a relevant exception is thrown.

File: examples/swig_purepython_boids_spatial3D/venv.bat [hunk 17]
```diff
@@ -0,0 +1,12 @@
...
+IF NOT "%1"=="" (
+    set build_dir=%1
+)
+set config=Debug
+IF NOT "%2"=="" (
+    set config=%2
+)
+cd ../../%build_dir%/lib/%config%/python/venv/Scripts
+call activate.bat
+cd ../../../../../../examples/swig_boids_spatial3D
```
[the remark this answers is not in the document]
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/venv.bat:12]: These are copied from the existing example. I am not sure why they are needed as they are windows specific so can probably just go in the bin.

File: swig/python/codegen/codegen.py (hunk 7 above)
Comment (mondus) [reply swig/python/codegen/codegen.py:8]: Done. Also removed from `setup.py.in`.

File: swig/python/codegen/codegen.py (hunk 9 above)
Comment (mondus) [reply swig/python/codegen/codegen.py:545]: manually fixed these

File: swig/python/codegen/codegen.py (hunk 12 above)
Comment (mondus) [reply swig/python/codegen/codegen.py:40]: Because math.h does not have a definition for `TAU`

File: swig/python/CMakeLists.txt (hunk 2 above)
Comment (ptheywood) [reply swig/python/CMakeLists.txt:240]: see https://github.com/FLAMEGPU/FLAMEGPU2/pull/887

File: swig/python/CMakeLists.txt (hunk 3 above)
Comment (ptheywood) [reply swig/python/CMakeLists.txt:295]: maybe resolved by https://github.com/FLAMEGPU/FLAMEGPU2/pull/887

File: swig/python/codegen/codegen.py (hunk 12 above)
Comment (ptheywood) [reply swig/python/codegen/codegen.py:40]: Fair, might be worth adding to the docs then that `math.tau` is not supprted, but can be computed as `2 * math.pi`

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 18]
```diff
@@ -0,0 +1,411 @@
...
+    pyflamegpu.setVariableFloat("y", agent_y)
+    pyflamegpu.setVariableFloat("z", agent_z)
+
+    pyflamegpu.setVariableFloat("fx", agent_fx)
+    pyflamegpu.setVariableFloat("fy", agent_fy)
+    pyflamegpu.setVariableFloat("fz", agent_fz)
+
+    return pyflamegpu.ALIVE
+
+    return pyflamegpu.ALIVE
```
[the remark this answers is not in the document]
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/boids_spatial3D.py]: Fixed

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 19]
```diff
@@ -0,0 +1,411 @@
+
+from textwrap import indent
+from pyflamegpu import *
+import pyflamegpu.codegen
+#from pyflamegpu import *
```
[the remark this answers is not in the document]
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/boids_spatial3D.py]: removed

File: swig/python/codegen/__init__.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (mondus) [reply swig/python/codegen/__init__.py]: Have removed this test as you are right its not required.

File: swig/python/codegen/__init__.py [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+from six.moves import cStringIO
+from .codegen import CodeGenerator
+from .codegen import CodeGenException
+from typing import Callable, Union
+import ast
+import inspect
+
+def codegen(tree: ast.AST) -> str:
+    """
+    Code generate function will tranlate a pure python FLAME GPU function from an ast tree
```
[the remark this answers is not in the document]
Comment (mondus) [reply swig/python/codegen/__init__.py]: fixed (along with a  few others that were spotted). VSCode spell is nice.

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 21]
```diff
@@ -0,0 +1,411 @@
...
+import sys, random, math
+
+"""
+  Pure python version of inputdata agent function for Boid agents, which reads data from neighboring Boid agents, to perform the boid flocking model.
+"""
+
+# Vector utility functions, see top of file for versions with commentary
+
+"""
+  Multiply a vector by a scalar value in-place
```
[the remark this answers is not in the document]
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/boids_spatial3D.py]: Fixed. See my previous comment on supporting this within device and agent functions.

File: examples/swig_purepython_boids_spatial3D/venv.bat (hunk 17 above)
[the remark this answers is not in the document]
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/venv.bat:12]: Have left this as we either have it consistent with the original python example or remove from both.

File: swig/python/codegen/__init__.py (hunk 4 above)
Comment (mondus) [reply swig/python/codegen/__init__.py]: Killed all references to six

File: swig/python/setup.py.in (hunk 5 above)
Comment (mondus) [reply swig/python/setup.py.in]: Gone

File: swig/python/codegen/codegen.py (hunk 12 above)
Comment (mondus) [reply swig/python/codegen/codegen.py:38]: Will do.

**Comment by Robadob** [comment]
> Add some documentation and clarify that this is an experimental feature

Will this include all agent fn code examples having a "PyScript" equivalent added?

**Comment by Robadob** [comment]
Is it able to convert

```py
print(`%f: %d`%(foo, bar))
```
to
```c++
printf("%f: %d\n", foo, bar);
```

It's a simple, but unusual transform, so not obvious whether it would be supported.

**Comment by mondus** [comment]
@Robadob To answer your questions. No it wont translate print functions. It will give a warning that the print function is not a flamegpu device agent or built in flamegpu function. It will then do its best to translate it as a general function call. E.g. remove the `;`. If you wanted to print in the Python agent function you would have to call `printf("%f: %d\n", foo, bar)`.

In terms of docs then yes I will update all the agent fn code examples.

File: swig/python/CMakeLists.txt (hunk 2 above)
Comment (mondus) [reply swig/python/CMakeLists.txt]: Resolved by PR #887 now merged int this branch

File: swig/python/CMakeLists.txt (hunk 3 above)
Comment (mondus) [reply swig/python/CMakeLists.txt]: Resolved by PR #887 now merged into this branch

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 22]
```diff
@@ -0,0 +1,404 @@
+
```
Comment (ptheywood) [inline examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:1]:
```suggestion
!# /usr/bin/env python3
```

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 23]
```diff
@@ -0,0 +1,404 @@
...
+from textwrap import indent
+from pyflamegpu import *
+import pyflamegpu.codegen
+import sys, random, math
+
+
```
Comment (ptheywood) [inline examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:7]: Would benefit from a file-level docblock describing the file 

i.e. something like:

```suggestion
"""
  FLAME GPU 2 implementation of boids in python 3, using python to define agent functions rather than CUDA C++ RTC.
  This is based on the FLAME GPU 1 implementation, but with dynamic generation of agents.
  Agents are also clamped to be within the environment bounds, rather than wrapped as in FLAME GPU 1.
"""
```

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 24]
```diff
@@ -0,0 +1,404 @@
...
+
+# Vector utility functions, see top of file for versions with commentary
```
Comment (ptheywood) [inline examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:9]: > see top of file for versions with commentary

This is the top of the file?

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 25]
```diff
@@ -0,0 +1,404 @@
...
+      @param z z component of the vector
+    """
+    # Get the length (dont call the device vec3Length func)
+    length = math.sqrt(x * x + y * y + z * z)
+    vec3Div(x, y, z, length)
+
+# pyflamegpu agent and device functions
+
+@pyflamegpu.device_function
+def vec3Length(x: float, y: float, z: float) -> float :
```
Comment (ptheywood) [inline examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:49]: Docblock missing

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 26]
```diff
@@ -0,0 +1,404 @@
...
+    return math.sqrtf(x * x + y * y + z * z)
+
+@pyflamegpu.device_function
+def clamp(v : float, min: float, max: float) -> float:
```
Comment (ptheywood) [inline examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:53]: docblock missing

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py [hunk 27]
```diff
@@ -0,0 +1,404 @@
...
+env.newPropertyFloat("GLOBAL_SCALE", 0.15)
+
+# Rule scalers
+env.newPropertyFloat("STEER_SCALE", 0.055)
+env.newPropertyFloat("COLLISION_SCALE", 10.0)
+env.newPropertyFloat("MATCH_SCALE", 0.015)
+
+
+# Location message
+
```
Comment (ptheywood) [inline examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:295]:
```suggestion
```

File: README.md [hunk 28]
```diff
@@ -24,7 +24,7 @@ This includes a number of key Agent-based Modelling (ABM) building blocks such a
 ## Project Status
 <!-- Remove this section once it is no longer in pre-release / alpha state -->
 
-FLAME GPU 2 is currently in an pre-release (alpha) state, and although we hope there will not be significant changes to the API prior to a stable release there may be breaking changes as we fix issues, adjust the API and improve performance.
+FLAME GPU 2 is currently in an pre-release (alpha) state, and although we hope there will not be significant changes to the API prior to a stable release there may be breaking changes as we fix issues, adjust the API and improve performance. The use of *pure* Python agent functions (agent functiosn expressed as Python syntax which are transpiled to C++) is currently supported (see examples) but classes as an *experimental* feature.
```
Comment (ptheywood) [inline README.md:27]:
```suggestion
FLAME GPU 2 is currently in an pre-release (alpha) state, and although we hope there will not be significant changes to the API prior to a stable release there may be breaking changes as we fix issues, adjust the API and improve performance.
The use of *pure* Python agent functions (agent functions expressed as Python syntax which are transpiled to C++) is currently supported (see examples) but classed as an *experimental* feature.
```

Typo fixed above, no reason to not use a newline either. 

Not convinced by describing it as `currently supported` `but classed as an experimental feature`. Surely its just an experimental feature? Maybe:

```suggestion
FLAME GPU 2 is currently in an pre-release (alpha) state, and although we hope there will not be significant changes to the API prior to a stable release there may be breaking changes as we fix issues, adjust the API and improve performance.
The use of *pure* Python agent functions (agent functions expressed as Python syntax which are transpiled to C++) is currently experimental. See `examples/swig_purepython_boids/spatial3D` for an example.
```

I'm also not keen on it being called `pure python`. IMO we now have:

* CUDA C++
* CUDA C++ RTC
* Python 3 with CUDA C++ RTC agent functions
* Python 3 with (experimental, restricted subset of) Python 3 (and some c++ mixed in for good measure) agent functions)

**Review by ptheywood — CHANGES_REQUESTED** [review]
Previous feedback is all addressed.

Spotted a couple of trivial changes to the new example. 

Not keen on calling it `pure python`, see the comment about it. 

Renaming the example directory would be a good change too, but we'd need to decide on how we refer to the 2 python modes first really.

**Comment by mondus** [comment]
Note: codegen requires `setKey` function adding to https://github.com/FLAMEGPU/FLAMEGPU2/blob/10b183ada949a38c3cc0ec8bb3088689fb3c7dd2/swig/python/codegen/codegen.py#L85

**Comment by mondus** [comment]
function `at` is also required for message in at https://github.com/FLAMEGPU/FLAMEGPU2/blob/10b183ada949a38c3cc0ec8bb3088689fb3c7dd2/swig/python/codegen/codegen.py#L82

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py (hunk 24 above)
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/boids_spatial3D.py]: removed

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py (hunk 25 above)
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:51]: added

File: examples/swig_purepython_boids_spatial3D/boids_spatial3D.py (hunk 26 above)
Comment (mondus) [reply examples/swig_purepython_boids_spatial3D/boids_spatial3D.py:61]: added

File: README.md [hunk 29]
```diff
@@ -25,6 +25,7 @@ This includes a number of key Agent-based Modelling (ABM) building blocks such a
 <!-- Remove this section once it is no longer in pre-release / alpha state -->
 
 FLAME GPU 2 is currently in an pre-release (alpha) state, and although we hope there will not be significant changes to the API prior to a stable release there may be breaking changes as we fix issues, adjust the API and improve performance.
+The use of *pure* Python agent functions (agent functions expressed as Python syntax which are transpiled to C++) is currently supported (see examples) but classed as an *experimental* feature.
```
Comment (ptheywood) [inline README.md:28]: Still refers to `*pure* Python`, which has been removed from teh docs repo.

**Review by ptheywood — CHANGES_REQUESTED** [review]
One occurance of `*pure* python` in some markdown, plus still in the example direcotry name `examples/swig_purepython_boids_spatial3D`. 

Other outstanding points all addressed / defferred to issues.

File: README.md (hunk 29 above)
Comment (mondus) [reply README.md]: ok fixed

**Review by ptheywood — APPROVED** [review]
You've still not renamed the example dir to remove `pure`. 

I'll mark it approved though and we can change it later given moving examples is a planned change, though I'd prefer if it were done.

**Comment by mondus** [comment]
Have renamed the example. Can merge tomorrow after CI has completed.

**Review by MILeach — APPROVED** [review]
